drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+		u32_t to_read = max(I2C_ESP32_BUFFER_SIZE, msg.len - 1);
+
+		/* Might be the last byte, in which case, `to_read` will
+		 * be 0 here.  See comment below.
+		 */
+		if (to_read) {
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_READ,
+				.num_bytes = to_read,
+			};
+		}
+
+		/* I2C master won't acknowledge the last byte read from the
+		 * slave device.  Divide the read command in two segments as
+		 * recommended by the ESP32 Technical Reference Manual.
+		 */
+		if (msg.len - to_read > 1) {
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_END
+			};
+		} else {
+			/* Read the last byte and explicitly ask for an
+			 * acknowledgement.
+			 */
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_READ,
+				.num_bytes = 1,
+				.ack_val = true,
+			};
+
+			/* Account for the `msg.len - 1` when clamping
+			 * transmission length to FIFO buffer size.
+			 */
+			to_read++;
+
+			if (msg.flags & I2C_MSG_STOP) {
+				wait_cmd = cmd;
+				*cmd++ = (struct i2c_esp32_cmd) {
+					.opcode = I2C_ESP32_OP_STOP
+				};
+			}
+		}
+
+		r = i2c_esp32_transmit_and_wait(dev, wait_cmd);
+		if (r < 0) {
+			return r;
+		}
+
+		for (i = 0; i < to_read; i++) {
+			*msg.buf++ = *config->regs.fifo & I2C_FIFO_RDATA;
+		}
+		msg.len -= to_read;
+
+		i2c_esp32_reset_fifo(config);
+		cmd_base = cmd_after_write;
+	}
+
+	return 0;
+}
+
+static int i2c_esp32_write_msg(struct device *dev, u16_t addr,
+			       struct i2c_msg msg)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	u32_t addr_len = 1 + data->config.bits.use_10_bit_addr;
+	bool first = true;
+
+	for (msg.len += addr_len; msg.len; ) {
+		volatile struct i2c_esp32_cmd *cmd =
+			(struct i2c_esp32_cmd *)config->regs.cmd_base;
+		u32_t to_send = min(I2C_ESP32_BUFFER_SIZE, msg.len);
+		size_t rem_cmds;
+		u32_t i;
+		bool send_end;
+		int r;
+
+		i2c_esp32_reset_fifo(config);
+		i2c_esp32_ack_intrs(config);
+
+		if (msg.flags & I2C_MSG_RESTART) {
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_RSTART
+			};
+		}
+
+		if (first) {
+			msg.len -= addr_len;
+			to_send -= addr_len;
+			first = false;
+
+			/* Send address */
+			*config->regs.fifo = addr & I2C_FIFO_RDATA;
+			if (data->config.bits.use_10_bit_addr) {
+				*config->regs.fifo = (addr >> 8) & I2C_FIFO_RDATA;
+			}
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_WRITE,
+				.num_bytes = addr_len,
+			};
+		}
+
+		/* Copy data to TX fifo */
+		for (i = 0; i < to_send; i++) {
+			*config->regs.fifo = *msg.buf++;
+		}
+		*cmd++ = (struct i2c_esp32_cmd) {
+			.opcode = I2C_ESP32_OP_WRITE,
+			.num_bytes = to_send,
+			.ack_en = true,
+		};
+		msg.len -= to_send;
+
+		send_end = msg.len && !(msg.flags & I2C_MSG_STOP);
+		*cmd++ = (struct i2c_esp32_cmd) {
+			.opcode = send_end ?
+				I2C_ESP32_OP_END : I2C_ESP32_OP_STOP,
+		};
+
+		/* Ensure the remaining I2C commands are cleared */
```

Comment:
Why?  Isn't there a "command count" register or something similar?  Or the hardware really wants to read out zero-valued commands as "noops"?

Reply:
This is done by the FreeRTOS driver as well.  I don't know if this is strictly necessary (i.e. the controller will stop when an END or a STOP opcode is found).